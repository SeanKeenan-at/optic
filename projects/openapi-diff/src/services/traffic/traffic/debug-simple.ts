import { ApiTraffic } from '../types';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';

export class SimpleExampleTraffic implements ApiTraffic {
  constructor(
    private _request: { method: string; path: string },
    private _response: { statusCode: string; json?: any }
  ) {}
  get method() {
    return this._request.method.toLowerCase() as OpenAPIV3.HttpMethods;
  }
  get path() {
    return this._request.path;
  }
  get response() {
    return {
      statusCode: this._response.statusCode,
      body: {
        contentType: this._response.json ? 'application/json' : undefined,
        jsonBodyString: this._response.json
          ? JSON.stringify(this._response.json)
          : undefined,
      },
    };
  }
}

export function makeExample(
  path: string,
  method: string | OpenAPIV3.HttpMethods,
  statusCode: string = '200',
  jsonBody?: object | undefined
) {
  return new SimpleExampleTraffic(
    { path, method },
    { statusCode, json: jsonBody }
  );
}
