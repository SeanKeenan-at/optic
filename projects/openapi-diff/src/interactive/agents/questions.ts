/*
  Agents are presented with questions, and must decide how to answer them. Possible answers
  - see a new url | add a new path, ignore the path/method pattern
  - see a shape diff | extend spec, replace, type
  - see a new body   | add it, ignore it
 */

import { ApiTraffic } from '../../services/traffic/types';
import { DiffType, IDiff, UnmatchedPath } from '../../services/diff/types';
import { v4 as uuidv4 } from 'uuid';

export type QuestionsForAgent = AddPathQuestion;

export type QuestionsForAgentAnswered = QuestionsForAgent & {
  answer: QuestionsForAgent['answer'];
};

interface Questions<A> {
  type: AnswerQuestionTypes;
  answer?: A;
  uuid: string;
}

export enum AnswerQuestionTypes {
  AddPath,
}

// Path add  Q / A
export interface AddPathQuestion extends Questions<AddPathAnswer> {
  type: AnswerQuestionTypes.AddPath;
  diff: UnmatchedPath;
  example: ApiTraffic;
}
export interface AddPathAnswer {
  pathPattern: string;
}
