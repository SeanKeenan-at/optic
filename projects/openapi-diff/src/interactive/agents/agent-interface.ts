import { AnswerQuestionTypes, QuestionsForAgent } from './questions';
import { ApiTraffic } from '../../services/traffic/types';
import { IDiff, IDiffService } from '../../services/diff/types';
import { IPatchOpenAPI } from '../../services/patch/types';

export interface AgentContext {}

export enum AgentEventEnum {
  DiffReceived = 'DiffReceived',
  Reset = 'Reset',
  AnswerQuestion = 'AnswerQuestion',
  SkipInteraction = 'SkipInteraction',
  SkipQuestion = 'SkipQuestion',
  AdvancePastError = 'AdvancePastError',
  Start = 'Start',
  Finish = 'Finish',
}

export type ProcessDiffContext = AgentContext & {
  diffs: IDiff[];
  traffic: ApiTraffic;
};
export type WaitingOnInputDiffContext = AgentContext & {
  diffs: IDiff[];
  questions: QuestionsForAgent[];
  traffic: ApiTraffic;
};

export type AgentEvent =
  | { type: AgentEventEnum.DiffReceived; diffs: IDiff[]; traffic: ApiTraffic }
  | { type: AgentEventEnum.AnswerQuestion; id: string; answer: any }
  | { type: AgentEventEnum.Reset }
  | { type: AgentEventEnum.SkipInteraction }
  | { type: AgentEventEnum.SkipQuestion };

export type AgentTypestate =
  | {
      value: 'idle';
      context: AgentContext;
    }
  | {
      value: 'process_diffs';
      context: ProcessDiffContext;
    }
  | {
      value: 'waiting_for_input';
      context: WaitingOnInputDiffContext;
    }
  | {
      value: 'error';
      context: AgentContext & { error: string };
    }
  | {
      value: 'finished';
      context: AgentContext & {};
    };

export interface AgentIntent {
  name: string;
  handleDiffs: (
    diff: IDiff,
    example: ApiTraffic,
    patch: IPatchOpenAPI,
    diffService: IDiffService,
    askQuestion: (question: QuestionsForAgent) => void
  ) => void;
  applyAnswerAsPatch: (
    questionAnswer: QuestionsForAgent,
    patch: IPatchOpenAPI
  ) => void;
}
