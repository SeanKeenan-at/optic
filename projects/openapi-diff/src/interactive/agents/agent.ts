import { assign, createMachine, interpret } from 'xstate';
import {
  AgentContext,
  AgentEvent,
  AgentEventEnum,
  AgentIntent,
  AgentTypestate,
  ProcessDiffContext,
  WaitingOnInputDiffContext,
} from './agent-interface';
import isEqual from 'lodash.isequal';
import { InteractiveDiffMachineType } from '../machine';
import { IDiff } from '../../services/diff/types';
import { QuestionsForAgent } from './questions';
import { ApiTraffic } from '../../services/traffic/types';
import { DiffEventEnum } from '../machine-interface';

export function createAgentMachine(
  diffMachine: InteractiveDiffMachineType,
  intent: AgentIntent
) {
  const agentMachine = createMachine<AgentContext, AgentEvent, AgentTypestate>({
    id: 'agent',
    initial: 'idle',
    context: {},
    states: {
      idle: {
        entry: assign({}),
        on: {
          [AgentEventEnum.DiffReceived]: {
            target: 'process_diffs',
            actions: assign((context, event) => {
              return {
                diffs: event.diffs,
                traffic: event.traffic,
                questions: [],
              };
            }),
          },
        },
      },
      process_diffs: {
        invoke: {
          // this entire src could be made a service for other kinds of diff intents
          src: async (
            ctx,
            event
          ): Promise<{ questions: QuestionsForAgent[] }> => {
            const { diffs, traffic } = ctx as ProcessDiffContext;

            // for user input
            const questions: QuestionsForAgent[] = [];
            // use this to patch diffs this intent automates
            const { specInterface, diffService } = diffMachine.state.context;

            diffs.forEach((diff) =>
              intent.handleDiffs(
                diff,
                traffic,
                specInterface.patch,
                diffService,
                (question) => {
                  questions.push(question);
                }
              )
            );

            // if automated patches are dispatched, we run these first
            // if (specInterface.patch.listPatches().length) {
            //   // diffMachine.send({
            //   //   type: DiffEventEnum.Agent_Submitted_Patch,
            //   // });
            //   return { questions: [] };
            // }

            return { questions };
          },
          onDone: [
            {
              target: 'check_should_flush',
              actions: assign((ctx, event) => {
                return {
                  questions: event.data.questions,
                };
              }),
            },
          ],
          onError: {
            actions: (ctx, event) => {
              console.log(event);
            },
          },
        },
      },
      waiting_for_input: {
        on: {
          [AgentEventEnum.SkipInteraction]: {
            actions: (ctx, event) => {
              diffMachine.send({
                type: DiffEventEnum.Agent_Skipped_Interaction,
              });
            },
            target: 'idle',
          },
          [AgentEventEnum.AnswerQuestion]: {
            actions: assign((ctx, event) => {
              const context = ctx as WaitingOnInputDiffContext;
              const { specInterface } = diffMachine.state.context;
              const question = context.questions.find(
                (i) => i.uuid === event.id
              );
              if (question) {
                question.answer = event.answer;
              }
              return { questions: context.questions };
            }),
            target: 'check_should_flush',
          },
        },
      },
      check_should_flush: {
        invoke: {
          src: async (ctx) => {
            const context = ctx as WaitingOnInputDiffContext;
            const { specInterface } = diffMachine.state.context;

            const result = { shouldIdle: true, shouldFlush: false };

            if (
              context.questions.length > 0 &&
              context.questions.every((i) => i.answer)
            ) {
              context.questions.forEach((question) =>
                intent.applyAnswerAsPatch(question, specInterface.patch)
              );
              result.shouldFlush = true;
            }

            if (
              context.questions.length > 0 &&
              context.questions.some((i) => !i.answer)
            ) {
              result.shouldFlush = false;
              result.shouldIdle = false;
            }

            if (
              context.questions.length === 0 &&
              specInterface.patch.listPatches().length > 0
            ) {
              result.shouldFlush = true;
            }

            return result;
          },
          onDone: [
            {
              target: 'idle',
              actions: () => {
                diffMachine.send({ type: DiffEventEnum.Agent_Submitted_Patch });
              },
              cond: (ctx, event) => event.data.shouldFlush,
            },
            {
              target: 'idle',
              cond: (ctx, event) => event.data.shouldIdle,
            },
            { target: 'waiting_for_input' },
          ],
          onError: {
            actions: [
              (ctx, error) => {
                console.error(error);
                diffMachine.send({
                  type: DiffEventEnum.Agent_Skipped_Interaction,
                });
              },
            ],
            target: 'idle',
          },
        },
      },
    },
  });

  const service = interpret(agentMachine);

  service.start();

  // response to state changes of diff machine
  let previousAsk: { diffs: IDiff[]; traffic: ApiTraffic } | undefined;

  diffMachine.onTransition((state, event) => {
    // transitioned to waiting_for_input

    if (state.value === 'waiting_for_input') {
      if (
        !isEqual(previousAsk, {
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        })
      ) {
        service.send({
          type: AgentEventEnum.DiffReceived,
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        });
        previousAsk = {
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        };
      }
    } else {
      previousAsk = undefined;
    }
  });

  return service;
}
