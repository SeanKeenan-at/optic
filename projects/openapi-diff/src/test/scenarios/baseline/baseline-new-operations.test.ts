import { AnswerQuestionTypes } from '../../../interactive/agents/questions';
import { emptySpec, oneEndpointExampleSpec } from '../specs';
import { DebugTraffic } from '../traffic';

describe('baseline new operations', () => {
  it('asks question for a new path', async () => {
    const scenario = emptySpec();
    scenario.sendTraffic(
      DebugTraffic('get', '/example').withJsonResponse({ hello: 'world' })
    );

    await scenario.answerQuestion(AnswerQuestionTypes.AddPath, (question) => {
      question.answer = {
        pathPattern: '/example',
      };
    });

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new http methods on matched paths', async () => {
    const scenario = await oneEndpointExampleSpec();

    scenario.sendTraffic(
      DebugTraffic('put', '/example').withJsonResponse({ hello: 'world' })
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new properties in matched bodies', async () => {
    const scenario = await oneEndpointExampleSpec();

    scenario.sendTraffic(
      DebugTraffic('get', '/example').withJsonResponse({
        hello: 'world',
        goodbye: 'saturn',
      })
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new query parameters on matched endpoints', async () => {
    const scenario = await oneEndpointExampleSpec();

    scenario.sendTraffic(
      DebugTraffic('get', '/example')
        .withJsonResponse({ hello: 'world' })
        .withQuery('limit=50')
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new response status codes on matched paths', async () => {
    const scenario = await oneEndpointExampleSpec();

    scenario.sendTraffic(
      DebugTraffic('get', '/example')
        .withStatusCode('404')
        .withJsonResponse({ error: 'look here...' })
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });
});
