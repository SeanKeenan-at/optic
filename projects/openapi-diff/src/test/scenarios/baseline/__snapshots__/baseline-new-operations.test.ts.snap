// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseline new operations asks question for a new path 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline new operations autolearns new http methods on matched paths 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
      "put": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
  },
}
`;

exports[`baseline new operations autolearns new properties in matched bodies 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "goodbye": Object {
                      "type": "string",
                    },
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                    "goodbye",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline new operations autolearns new query parameters on matched endpoints 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": Object {
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline new operations autolearns new response status codes on matched paths 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
          "404": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "error": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "error",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;
